### ISSUE TITLE
Use with SFTP

### ISSUE THREAD

**Issue by mu-arch** [netheril96/securefs#158]
How can I use this with a remote location mounted via SSHFS? Mounting the folder via SSH in terminal causes SecureFS to print errors and not work.

**Comment by netheril96** [comment]
What errors do you see?

**Comment by mu-arch** [comment]
Thanks for the quick response. The errors look like below, I've redacted my personal info with @ signs.

```
[Error] [0x7f360d369640] [2023-07-31 22:06:25.924213993 UTC] [int securefs::operations::opendir(const char*, fuse_file_info*):331]    Function fails with arguments ("/", {fh=0x(nil), flags=0304000}) with return code -95 because it encounters exception securefs::POSIXException: Operation not supported (Opening /run/user/1000/gvfs/sftp:host=@@@@@@@@@@,user=@@@@@@@@@@@@@@@@@@@@@@@@/00/00000000000000000000000000000000000000000000000000000000000000.meta with flags 02)
```

The errors spam for a moment and then the program exits. I saw numerous operations fail such as opening files, reading file attributes, etc.

**Comment by mu-arch** [comment]
Correction, in my example I used SFTP to connect. However, I have been unable to connect with SSHFS either, though this may be a misconfig on my end. I'd like to be able to use it with SFTP if possible though.

**Comment by netheril96** [comment]
Are you able to mount an another FUSE on top of SFTP/SSHFS?

**Comment by mu-arch** [comment]
> Are you able to mount an another FUSE on top of SFTP/SSHFS?

I haven't tried. My use case is a Linux based computer connecting to a Mac and storing securefs files in iCloud. I never plan to read the files on the Mac, just the Linux computer.

**Comment by mu-arch** [comment]
So I tried SMB also but it just seems it's gvfs no matter if I use SFTP, or SMB causing those problems with unsupported commands. SSHFS seemed the most promising but I can't figure out a permissions issue.

**Comment by netheril96** [comment]
I just tried mounting `securefs` over a `sshfs` volume on Linux Mint. It works without any problems.

You need to give me more details on how to reproduce the issue.

**Comment by netheril96** [comment]
Can you try `fuse-zip` over SFTP? I suspect that GVFS is the culprit here.

**Comment by mu-arch** [comment]
Excuse my ignorance here, but how does fuse-zip equate to GVFS? Isn't fuse-zip for mounting a archive as if it were a folder on the filesystem, while GVFS mounts a remote location (one of it's many features) as if it were a local device?

**Comment by netheril96** [comment]
Let me rephrase.

I suspect that mounting *any* FUSE filesystem on top of GVFS won't work. So I ask you to try mounting a `fuse-zip` over GVFS. If it indeed does not work, then the bug is in GVFS and you should report it to them.

If `fuse-zip` over GVFS works but `securefs` over GVFS doesn't, then it is a bug of `securefs` and I will investigate if I can fix it.

**Comment by mu-arch** [comment]
I never tried what you said because I got it working with SSHFS but I have another related question.

I'm getting error spam when I upload files in full and lite modes.

In full mode there are no errors until I upload my 124th file. After that any file uploaded fails with the message "Operation not permitted".

```
[Error] [0x7f7852d23640] [2023-08-14 16:46:26.734260753 UTC] [int securefs::operations::create(const char*, mode_t, fuse_file_info*):420]    Function fails with arguments ("/myfolder/file.jpg", 33261, {fh=0x(nil), flags=0100301}) with return code -1 because it encounters exception securefs::POSIXException: Operation not permitted (Opening /home/myuser/Desktop/mac_remote_sshfs/mount_point/Documents/Random/archive/24/7d335873dfe07823e168c3347y726fa3574e9ca35bcb912a1543fac79dfeba.meta with flags 0302)
```

In lite mode every file appears to upload fine, and the problem with the 124th file does not exist, however I receive an error spam of "Operation not permitted" from the console.

```
[Error] [0x7f574ffff640] [2023-08-14 17:04:30.098153283 UTC] [int securefs::lite::chown(const char*, uid_t, gid_t):403]    Function fails with arguments ("/myfolder", 1000, 1000) with return code -13 because it encounters exception securefs::POSIXException: Permission denied (chown /home/myuser/Desktop/mac_remote_sshfs/mount_point/Documents/Random/archive/Z4E3FNYRU6MBSN5KBCRCCYTFKZ7CAT3SUYV7KTIMZUISMXY9 with uid=1000 and gid=1000)
```

Let me remind you of my setup, as I read about some issues with xattr on Mac in the design doc. I am using SecureFS on my Linux computer with a remote Mac mounted via SSHFS. I store the SecureFS folder on the remote Mac and access it with the SecureFS client on the Linux computer. Perhaps SecureFS is writing files as it would on a Linux FS rather than using your shims for making it work on Mac, since it believes it's on Linux?

**Comment by netheril96** [comment]
> In full mode there are no errors until I upload my 124th file. After that any file uploaded fails with the message "Operation not permitted".

It's because macOS has very low number of allowed open file descriptors. If you run `securefs` on a mac, it raises that automatically. But on SSHFS it cannot do it.

On your mac, run this `launchctl limit maxfiles 8192 unlimited` (it doesn't persist across reboot) before you mount the sshfs volume. Or you can run this with ssh, exit it, and then mount sshfs.

> In lite mode every file appears to upload fine, and the problem with the 124th file does not exist, however I receive an error spam of "Operation not permitted" from the console.

This is because FUSE and SSHFS in particular all have problems mapping user ids. If it doesn't affect your usage, you can ignore these logs.

**Comment by mu-arch** [comment]
With your permission, I could create a new file under the docs folder to document this.
